### ISSUE TITLE
Decaying gradient issue 

### ISSUE THREAD

**Issue by Weinmarr** [rusty1s/pytorch_geometric#2020]
Hi Matthias,

I am using GCNConv to solve a prediction task problem with linear layers at the output of the GNN. The model is trained on graphs of 10K nodes with ~20K-40K edges. The gradient value decays down to 0 after less than 10 epochs when training on these graphs and has an overall bad gradient flow during backpropagation. For reference, I have attached the code:

```python
import torch
from torch.nn.parameter import Parameter
import torch.nn as nn
from torch_geometric.nn import GCNConv


class Model(nn.Module):
    def __init__(self, nin, nhid1, nout, hid_l, out_l=1):
        super(Model, self).__init__()
       
        self.gc1 = GCNConv(in_channels= nin, out_channels= nhid1)
        self.gc2 = GCNConv(in_channels= nhid1, out_channels= nout)
        self.lay1 = nn.Linear(nout ,hid_l)
        self.l0 = nn.Linear(hid_l,hid_l)
        self.l1 = nn.Linear(hid_l,hid_l)
        self.lay2 = nn.Linear(hid_l ,out_l)
        self.active = nn.LeakyReLU(0.1)
       
        with torch.no_grad():
            self.gc1.weight = Parameter(nn.init.uniform_(torch.empty(nin,nhid1),a=0.0,b=1.0))
            self.gc1.bias = Parameter(nn.init.uniform_(torch.empty(nhid1),a=0.0,b=1.0))
            self.gc2.weight = Parameter(nn.init.uniform_(torch.empty(nhid1,nout),a=0.0,b=1.0))
            self.gc2.bias = Parameter(nn.init.uniform_(torch.empty(nout),a=0.0,b=1.0))
            self.lay1.weight = Parameter(nn.init.uniform_(torch.empty(hid_l, nout ),a=0.0,b=1.0))
            self.l0.weight = Parameter(nn.init.uniform_(torch.empty(hid_l, hid_l),a=0.0,b=1.0))
            self.l1.weight = Parameter(nn.init.uniform_(torch.empty(hid_l, hid_l),a=0.0,b=1.0))
            self.lay2.weight = Parameter(nn.init.uniform_(torch.empty(out_l,hid_l),a=0.0,b=1.0))
                       

    def forward(self, data):
        x, adj = data.x, data.edge_index
        x = self.active(self.gc1(x, adj))
        x = self.active(self.gc2(x, adj))
        x = self.active(self.lay1(x))
        x = self.active(self.l0(x))
        x = self.active(self.l1(x))
        x = self.active(self.lay2(x))
       
        return x
```

The gradient flow for the model is: 

![Figure_1](https://user-images.githubusercontent.com/19278722/104810327-ba2a3400-582e-11eb-9415-757393b1fee1.png)

As it can be seen, the gradient falls steeply after the second GCN layer and doesn't improve upon later. Is there a way to improve the gradient values while training on these graphs?

Thanks for the help and wonderful library.

**Comment by Weinmarr** [comment]
@rusty1s  It is a node regression problem that outputs a [n X 1] vector (each entry in the vector corresponds to a node). Removing the last activation didn't change anything and the gradient flow remains the same. Is the issue arising because of Batch Gradient Descent as here:    https://discuss.pytorch.org/t/very-small-lstm-gradients/59234/2?u=asgard

**Comment by rusty1s** [comment]
It might be possible that full-batch training leads to smaller gradient due to averaging, but I never noticed that in practice. You can try to add `BatchNorm` after each intermediate `GCNConv` layer to obtain better gradients.
